### PR TITLE
Add instance methods for javascript processor so they can be extended

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
I'm developing a plugin to obtain JS sourcemaps from private sources which require authentication (like S3). I need to override the implementation of both `discover_sourcemap` and `fetch_sourcemap`.

### CODE REVIEW

**Comment by dcramer** [comment]
I'm not super excited about the idea of fragmenting this code as it will be harder to maintain compatibility. Ideally you'd use the releases API so everything stays standard. Is there a good reason not to?

**Comment by ChadKillingsworth** [comment]
Mainly I wanted to add fallback methods such that the calls in a plugin look like:

```Python
soucemap = super(SourceProcessor, self).discover_sourcemap(result)
if not sourcemap
    //do custom work here
```

I might be able to use the releases api to do this, but I haven't found documentation that even describe it (besides setting a release version in a client).

I'm having a hard time understanding why this would fragment the code though. It simply gives a way to augment the functionality of these methods. I have no desire to wholesale rewrite them. Without this change I'm left with rewriting all of the `populate_source_cache` method - which I'd really like to avoid.

**Comment by dcramer** [comment]
@ChadKillingsworth what I mean by the releases API is the artifact storage. We (quietly) added support for uploading sourcemaps (and effectively source) to Sentry.

I guess fragment isnt the appropriate word. Mostly I'm concerned that "we'll likely change this a lot" so creating a public API around it is could cause us a bunch of headaches.

I'm willing to consider it if we determine the releases' stored artifacts solution isn't a good enough approach here, but I'd like to avoid it if possible.

(Another alternative would be that we make release storage pluggable somehow which would probably also be a better solution, but we haven't abstracted release storage yet and it's still hardcoded in the JS stuff)

**Comment by ChadKillingsworth** [comment]
The releases API will handle most use cases (assuming it works the way I think it does). However there's still one that isn't covered.

I'm investigating the option of having the web server auto-wrap javascript in try/catch blocks where the catch is an extremely lightweight js client. In this scenario, the sentry client won't have the release version of the actual code. This is how Google servers operate.

I'm willing to approach the problem in a different way. Ideally we could find a method that's usable in getsentry.com as well.

File: src/sentry/lang/javascript/processor.py [hunk 1]
```diff
@@ -572,3 +572,9 @@ def populate_source_cache(self, project, frames, release):
                         done_file_list.add(next_filename)
                     else:
                         pending_file_list.add(next_filename)
+
+    def discover_sourcemap(self, result):
+        return discover_sourcemap(result)
```
Comment (dcramer) [inline src/sentry/lang/javascript/processor.py]: afaik this should cause recursion issues in Python

Granted I'm only half-awake so maybe I'm just not seeing something.

**Comment by dcramer** [comment]
@ChadKillingsworth I dont have a great answer beyond "Make the artifacts API pluggable" but its not clear to me if that solves your problem.

i.e. if you could provide a hook that can inject a response to "get the source file for FOO", would that effectively solve your problem?

I'd also be interested in hearing more about the try/catch/generation stuff. Is it primarily to catch on-load errors, or is it more than that? We're talking about creating a build tool which will actually inject try/catch on basically 100% of your code, just to ensure we can get sane tracebacks.

**Comment by ChadKillingsworth** [comment]
I'd need two hooks:

 1. Get source file for FOO
 2. Sourcemap for file FOO (covers cases where the sourceMapUrl comment isn't present)

For the try/catch generation, yes it would only catch load/initialization errors. It doesn't solve the problem of asynch callbacks. I haven't decided whether to actually implement this or not. Frankly, with IE 10+, Chrome and FireFox now correctly exposing the stack to the window `error` event, I'm trying to decide if it's worth the effort. For my efforts, only Safari would lack a stacktrace.

**Comment by ChadKillingsworth** [comment]
I was able to use django-storages and the SENTRY_FILESTORE settings to handle most of this. However, I'd still like to pursue being able to override `discover_sourcemap` so that some files aren't required to have the source mapping url comment.

I'll update the PR to handle just that piece shortly.

**Pull request closed without merging**